src/odbc_connection.cpp

```diff
@@ -2782,108 +2843,340 @@ fetch_and_store
+              if (!SQL_SUCCEEDED(return_code))
+              {
+                return return_code;
+              } 
+
+              // If the data is null, simply indicate and continue to the next
+              // column
+              if (string_length_or_indicator == SQL_NULL_DATA)
+              {
+                row[column_index].size = SQL_NULL_DATA;
+                continue;
+              }
+              // If the data (+ whatever null terminator, not included in
+              // string_length_or_indicator) was larger than the size of the
+              // buffer, then need to call SQLGetData at least once more
+              else if (
+                string_length_or_indicator == SQL_NO_TOTAL || 
+                string_length_or_indicator >= buffer_size
+              ) 
+              {
+                // Continue looping as long as we don't know the final resize
+                // that we need. Once a buffer size is returned instead of
+                // SQL_NO_TOTAL, we still need to call SQLGetData one more time,
+                // and the break out of this loop.
+                bool last_get_data_call = false;
+                while (!last_get_data_call)
                 {
-                  SQLCHAR *memory_start = (SQLCHAR *)data->bound_columns[column_index].buffer + (row_index * data->columns[column_index]->buffer_size);
-                  row[column_index].size = strlen((const char *)memory_start);
-                  // Although fields going from SQL_C_CHAR to Napi::String use
-                  // row[column_index].size, NUMERIC data uses atof() which requires
-                  // a null terminator. Need to add an aditional byte.
-                  row[column_index].char_data = new SQLCHAR[row[column_index].size + 1]();
-                  memcpy
+                  // If not SQL_NO_TOTAL, we know how large to resize to buffer
+                  // to get the rest of the data.
+                  if (string_length_or_indicator != SQL_NO_TOTAL)
+                  {
+                    last_get_data_call = true;
+                  }
+
+                  // Handling the data that was returned depends heavily on the
+                  // target type of the binding data.
+                  switch(data->columns[column_index]->bind_type)
+                  {
+                    case SQL_C_BINARY:
+                      data_returned_length = buffer_size;
+                      buffer_size =
+                        string_length_or_indicator == SQL_NO_TOTAL ?
+                        buffer_size * 2 :
+                        row[column_index].size + string_length_or_indicator;
+                      row[column_index].char_data =
+                        (SQLCHAR *)
+                        realloc
+                        (
+                          row[column_index].char_data,
+                          buffer_size
+                        );
+                      row[column_index].size += data_returned_length;
```

Comment:
This is adding the data_returned_size after the size has already been increased above.

Reply:
`row[column_index].size` is intended to store the amount of _valid_ data inside the buffer. So for each call to `SQLGetData`, I determine how much valid data was returned for that calls. For `SQL_C_CHAR` and `SQL_C_WCHAR`, I use `strlen` and `strlen16`, but for `SQL_C_BINARY`, I assume that is equal to the `buffer_size` (which stores the size of the buffer starting at `target_pointer`). However looking at this logic, I think I am missing a key logic path: Calls to `SQL_NO_TOTAL` that then return a value `string_length_or_indicator` that is smaller than `buffer_size`. This logic always assume that one more call will be needed.

Reply:
Nevermind, I misread the code. We don't increase size above, we add size to the indicator and assign the result to buffer_size.